File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String str) {
           Stack<Character> st = new Stack<>();
        
                 for(int i=0; i<str.length(); i++){
                        char c = str.charAt(i);
                        if(c=='(' || c=='{' || c=='[' ) st.push(c);
                        else{
                                if(!st.empty() && isValid(st.peek())== c)  st.pop();
                                else  return false;                                
                        }
                 }
                
        return st.empty();
        
    }
    
     char isValid(char open){
            if(open=='(' )     return ')';
            if(open=='[')      return ']';                
            return '}';               
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isMatch(self,open,close):
        if open=='(' and close==')':
            return True
        if open=='[' and close==']':
            return True
        if open=='{' and close=='}':
            return True
        return False
    
    def isValid(self, s: str) -> bool:
        stack = []

        for ch in s:
            if ch in ['(','[','{']:
                stack.append(ch)
            elif len(stack)>0 and self.isMatch(stack[len(stack)-1],ch):
                stack.pop()
            else:
                return False

        return len(stack)==0
    
        
        